pkg/controllers/pod_controller.go

```diff
@@ -0,0 +1,182 @@
+
+	"sigs.k8s.io/jobset/pkg/util/collections"
+
+	jobset "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+)
+
+const (
+	podOwnerKey = ".metadata.controller"
+	nodePoolKey = "cloud.google.com/gke-nodepool"
+)
+
+var exclusiveGateConditionType = corev1.PodConditionType(jobset.ExclusiveGate)
+
+// PodReconciler reconciles a JobSet object
+type PodReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	Record record.EventRecorder
+}
+
+func NewPodReconciler(client client.Client, scheme *runtime.Scheme, record record.EventRecorder) *PodReconciler {
+	return &PodReconciler{Client: client, Scheme: scheme, Record: record}
+}
+
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;watch;update;patch
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch;create;update;patch;delete
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// Get JobSet from apiserver.
+	var pod corev1.Pod
+	if err := r.Get(ctx, req.NamespacedName, &pod); err != nil {
+		// we'll ignore not-found errors, since there is nothing we can do here.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// TODO: delete logging once we confirmed it works as it will be too noisy.
+	log := ctrl.LoggerFrom(ctx).WithValues("pod", klog.KObj(&pod))
+	ctx = ctrl.LoggerInto(ctx, log)
+	log.V(2).Info("Reconciling Pod")
+
+	// Check if this is the "leader" pod (completion index 0) and if it has been scheduled.
+	if pod.Annotations[batchv1.JobCompletionIndexAnnotation] == "0" && pod.Spec.NodeName != "" {
+		nodePool, err := r.nodePoolFromNode(ctx, pod.Spec.NodeName, pod.Namespace)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.assignGatedPodsToNodePool(ctx, &pod, nodePool); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PodReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Pod{}).
+		Complete(r)
+}
+
+func SetupPodReconcilerIndexes(ctx context.Context, indexer client.FieldIndexer) error {
+	return indexer.IndexField(ctx, &corev1.Pod{}, podOwnerKey, func(obj client.Object) []string {
+		o := obj.(*corev1.Pod)
+		owner := metav1.GetControllerOf(o)
+		if owner == nil {
+			return nil
+		}
+		// ...make sure it's a Pod owned by a Job...
+		if owner.Kind != "Job" || owner.APIVersion != batchv1.SchemeGroupVersion.String() {
+			return nil
+		}
+		return []string{owner.Name}
+	})
+}
+
+// nodePoolFromNode gets the node pool name that the given node is part of.
+func (r *PodReconciler) nodePoolFromNode(ctx context.Context, nodeName, ns string) (string, error) {
```

Comment:
yeah, we should have no reference to nodepools, lookup the topologyKey label value on the node.